t/cmd/lfs-ssh-echo.go

```diff
@@ -17,35 +20,78 @@ type sshResponse struct {
 	ExpiresIn int               `json:"expires_in,omitempty"`
 }
 
+func shell() string {
+	if runtime.GOOS == "windows" {
+		return "bash"
+	}
+	return "sh"
+}
+
+func spawnCommand(command string) error {
+	cmd := exec.Command(shell(), "-c", command)
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	err := cmd.Run()
+	if e, ok := err.(*exec.ExitError); ok {
+		var ws syscall.WaitStatus
+		ws, ok = e.ProcessState.Sys().(syscall.WaitStatus)
+		if ok {
+			os.Exit(ws.ExitStatus())
+		}
+	}
+	return err
+}
+
 func main() {
 	// expect args:
 	//   lfs-ssh-echo -p PORT -- git@127.0.0.1 git-lfs-authenticate REPO OPERATION
-	if len(os.Args) != 6 {
+	if len(os.Args) < 3 {
 		fmt.Fprintf(os.Stderr, "got %d args: %v", len(os.Args), os.Args)
 		os.Exit(1)
 	}
 
-	if os.Args[1] != "-p" {
-		fmt.Fprintf(os.Stderr, "$1 expected \"-p\", got %q", os.Args[1])
-		os.Exit(1)
+	offset := 4
+	if os.Args[1] == "--" {
+		offset = 2
+	} else if os.Args[1] == "git@127.0.0.1" {
+		offset = 1
+	} else {
+		if os.Args[1] != "-p" {
+			fmt.Fprintf(os.Stderr, "$1 expected \"-p\", got %q", os.Args[1])
+			os.Exit(1)
+		}
+
+		if os.Args[3] != "--" {
+			fmt.Fprintf(os.Stderr, "$3 expected \"--\", got %q", os.Args[3])
+			os.Exit(1)
+		}
 	}
 
-	if os.Args[3] != "--" {
-		fmt.Fprintf(os.Stderr, "$3 expected \"--\", got %q", os.Args[3])
+	if len(os.Args) < offset+2 {
+		fmt.Fprintf(os.Stderr, "got %d args: %v", len(os.Args), os.Args)
 		os.Exit(1)
 	}
 
-	if os.Args[4] != "git@127.0.0.1" {
-		fmt.Fprintf(os.Stderr, "$4 expected \"git@127.0.0.1\", got %q", os.Args[4])
+	if os.Args[offset] != "git@127.0.0.1" {
+		fmt.Fprintf(os.Stderr, "$4 expected \"git@127.0.0.1\", got %q", os.Args[offset])
 		os.Exit(1)
 	}
 
 	// just "git-lfs-authenticate REPO OPERATION"
-	authLine := strings.Split(os.Args[5], " ")
-	if len(authLine) < 13 {
+	authLine := strings.Split(os.Args[offset+1], " ")
+	if len(authLine) < 2 {
 		fmt.Fprintf(os.Stderr, "bad git-lfs-authenticate line: %s\nargs: %v", authLine, os.Args)
```

Comment:
I believe this needs to be `< 2` because if we invoke Git, there isn't any operation.  We get `git-receive-pack REPO` or `git-upload-pack REPO`.  Of course, if I had bothered to update the comment (which I have done since you suggested that), this would have been clearer, so my apologies for the oversight.

I think an `os.Exit(1)` is appropriate.  Having that before would have demonstrated that this was broken, but of course we didn't.  I'll explain this in the commit message as well.

Reply:
OK, now I think I understand what was confusing me before (along with the weird `< 13` check).  It's that the command and arguments to be executed remotely are expected to be wrapped in quotes, e.g., `lfs-ssh-echo git@127.0.0.1 "git-lfs-transfer REPO OPERATION"` or `lfs-ssh-echo -p PORT -- git@127.0.0.1 "git-lfs-authenticate REPO OPERATION"`.

That's because of this [line](https://github.com/git-lfs/git-lfs/blob/502801373173e28268b4f8417707117d4ee5fecd/lfshttp/ssh.go#L129) in what is presently `lfshttp.sshGetLFSExeAndArgs()` and this PR refactors to `ssh.GetLFSExeAndArgs()`, which packs all of the command and arguments to be executed remotely on the SSH server into a single argument to be passed to the local SSH client.

So, knowing that, I might suggest that the comments here and above still need a bit more revision, and the error message could also benefit from being made less specific to `git-lfs-authenticate`.  For the expanded comments above, just adding the quotes would help:
```go
	//   lfs-ssh-echo -p PORT -- git@127.0.0.1 "git-lfs-authenticate REPO OPERATION"
	//   lfs-ssh-echo -p PORT -- git@127.0.0.1 "git-lfs-transfer REPO OPERATION"
	//   lfs-ssh-echo -- git@127.0.0.1 "git-lfs-transfer REPO OPERATION"
	//   lfs-ssh-echo git@127.0.0.1 "git-upload-pack REPO"
	//   lfs-ssh-echo git@127.0.0.1 "git-receive-pack REPO"
```

And then for this section, maybe something like the following?
```go
	// just "COMMAND REPO OPERATION" (from the final quoted argument)
	authLine := strings.Split(os.Args[offset+1], " ")
	if len(authLine) < 2 {
		fmt.Fprintf(os.Stderr, "bad remote SSH command: %s\nargs: %v", authLine, os.Args)
		os.Exit(1)
	}
```

Oh, and it couldn't hurt to rename `authLine` to a more generic name, since it now covers remote commands other than just `git-lfs-authenticate`.